hathor/p2p/sync_checkpoints.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from itertools import chain
+from typing import TYPE_CHECKING, Dict, List, Optional, Tuple
+
+from structlog import get_logger
+
+from hathor.checkpoint import Checkpoint
+from hathor.transaction.block import Block
+from hathor.transaction.storage.exceptions import TransactionDoesNotExist
+
+if TYPE_CHECKING:
+    from hathor.manager import HathorManager  # noqa: F401
+    from hathor.p2p.node_sync_v2 import NodeBlockSync  # noqa: F401
+
+logger = get_logger()
+
+
+class SyncCheckpoint:
+    """ It is a central manager of the sync between checkpoints among all peers.
+    """
+
+    def __init__(self, manager: 'HathorManager'):
+        # All peers that have all the checkpoints to download
+        self.peers_to_request: List['NodeBlockSync'] = []
+
+        # All peers that we are connected but don't have all the checkpoints
+        self.incomplete_peers: List['NodeBlockSync'] = []
```

Comment:
Why do we use it for?

Reply:
All peers are remembered, except on separate lists, because when sync-checkpoint finished it will resume the "block sync" part of the sync-v2 on all of them.

Reply:
All peers are remembered, except on separate lists, because when sync-checkpoint finished it will resume the "block sync" part of the sync-v2 on all of them.